### DISCUSSION TITLE
Tagging bible chapters

### DISCUSSION THREAD

**Discussion by joey-kilgore** [tim-hub/obsidian-bible-reference#127] [Ideas]
From my current understanding of the plug-in, there is nothing within the project right now that allows for viewing how different notes may be linked (referencing the same books or chapters). I think there are 4 levels of tags that could be easily included (and I'll show an example of how for each):
1. A general `#bible` for all bible verses
2. A book specific tag (ex. `#matthew` for the Gospel of Matthew)
3. A chapter specific tag (ex. `#matthew26` for the 26th chapter in the Gospel of Matthew)
4. A verse specific tag (ex. `#matthew26_40` for the 40th verse in the 26th chapter in the Gospel of Matthew)

This could be done relatively unobtrusively on the same line that gives the link to the chapter online.
> [!Bible] [matthew 25:40 - ESV](https://bolls.life/ESV/40/25/) `#bible` `#matthew` `#matthew26` `#matthew26_40`
> <sup> 40 </sup>And the King will answer them, ‘Truly, I say to you, as you did it to one of the least of these my brothers,  you did it to me.’

While being able to search for verses might be sufficient, this inclusion would provide a simple solution to easily view connections of notes containing verses in the graph view (by turning on the tags in the graph view).

Each of these 4 options could be a selection item in the Bible Reference settings, and when turning them on and off should go and adjust the tags in all notes containing the bible verses (allowing for people to turn them on to whatever level makes sense for their search in the graph view).  
This last option of going back and modifying tags on all notes containing bible references might be something for later depending on the want from the community for this portion, but I believe the addition of the tags should be relatively straightforward (especially the `#bible`) while adding a sufficiently new and useful tool for seeing connected thoughts.

**Comment by JTravisWatson** [comment]
In my personal setup I use nested tags to denote references. For example, Romans 12:2 becomes #Romans/12/2 

These can be hidden in a comment block so they don’t clutter the note. I’m not too sure how multiple verses could be accounted for, unless you made a specific tag for a passage, such as #Genesis/1/1_3 for example.

Just a thought to give more options for formatting, though I’m not sure if that would make life easier on the backend or not.

**Comment by brennenputh** [comment]
Honestly, I think this feature is great and was definitely needed.  However, formatting is a bit of an issue.  It would be nice if, instead of editing the code to format a tag or link differently, the format could simply be in format-string form.  For example, a string of `{book}{chapter}` could output a tag such as `genesis1` (or a backlink of the same format).  Allowing a system like this would give much more flexibility and decrease the amount of options in the GUI for different styles.  I'm willing to make a PR to support this if desired.

**Comment by joey-kilgore** [comment]
@brennenputh This is really good suggestion. I had considered doing this, but I wanted to push an initial PR with the general idea to get it reviewed so we could have more of this discussion. If you would like to work on this and make a PR that would probably be worth starting, but one of the annoying things that still needs to be handled is standardizing book names (because sometimes the tag can be created using 'gen' instead of 'genesis'. Right now, there is some regex used for the book name, but I think we could create a much more robust standardization method in the utils to keep things more consistent. Unfortunately, this might be a bit trickier with the multi language support.

I'm wondering if there is a way to utilize some of the work already included in the [abbreviations.ts](https://github.com/tim-hub/obsidian-bible-reference/blob/master/src/data/abbreviations.ts) script to help with this, but I'm not sure.

**Comment by Xtremefaith** [comment]
I am so excited about this idea because I currently use https://github.com/Xtremefaith/BibleGateway-to-Obsidian which I forked and made adjustments so that users can easily setup an atomic note for every verse. 

Here is [a video](https://www.loom.com/share/d2d0b6b400b5447b8ce76374b57a443b) I made at the time to help explain the enhancements. Alternatively I've described them below

<details><summary>Detailed Explanation</summary>

This script allows users to generate their own personal digital bible using the [Bible Gateway](https://www.biblegateway.com/) API. There are several options involved with how a user might do this, but per my enhancements the following directory structure is created to allow for atomic **Verse Notes**.
<img width="690" alt="image" src="https://github.com/tim-hub/obsidian-bible-reference/assets/1228550/818df436-9a04-4181-bcd6-165e3920bc55">

Each **Chapter Note** then embeds all the necessary verse notes like this:

<img width="761" alt="image" src="https://github.com/tim-hub/obsidian-bible-reference/assets/1228550/cbeb7f98-86a9-4502-a1ae-50060d45f47e">

Additionally, it includes options to add features to the verse note. If you're an atomic note purest then you may not appreciate these specific features but you can choose to not include them when building your Personal Vault Bible. 

<img width="802" alt="image" src="https://github.com/tim-hub/obsidian-bible-reference/assets/1228550/1e3733b6-7c8f-45ae-bfd6-51fe448a16e3">

As you can see from this sample, I like this for the purpose of keeping specific interpretation notes on the verse rather than a generating unnamed notes to link. Also when reading through I can use the breadcrumbs. Again not for everyone. 

Therefore, there is a `verse` section link added when these features are added so that I can directly reference the verse without all the content of the note. For example, I often will link in my church/sermon notes with something like this `[[Genesis 1.1^verse|Genesis 1:1-3]]`
- Notice I can't actually link to all 3 verse with this method, but when reviewing my notes I know to read all 3 verses
- Also, my note only backlinks to the one verse referenced instead of all verse which again isn't ideal
- I could manually link all verse but determined that's too much effort

</details> 

---

Now with all this said, I think it would be so amazing if your "Create Book/Chapter" options worked with this approach (especially with the popularity around [BibleGateway-to-Obsidian](https://github.com/joebuhlig/BibleGateway-to-Obsidian) and the growing desire to have offline/local PKM tools. Technically the setup I described above has some great capabilities, but in the process of note making it's too difficult to leverage, but your plugin could change that.

Ideally, if I could type:

```
--Gen1:1-3
```

And your generator could produce the following output:

```
> [!Bible] [Gen 1:1-3 - ESV](https://bolls.life/ESV/1/1/) 
>  <sup> [[Genesis 1.1|1]] </sup>In the beginning, God created the heavens and the earth. <sup> [[Genesis 1.2|2]]  </sup>The earth was without form and void, and darkness was over the face of the deep. And the Spirit of God was hovering over the face of the waters. <sup> [[Genesis 1.3|3]]  </sup>And God said, “Let there be light,” and there was light.
```

Look how nice that looks!
<img width="620" alt="image" src="https://github.com/tim-hub/obsidian-bible-reference/assets/1228550/9e744bfe-412f-4782-86a2-9aa6b4265331">

This now has the following benefits:
1. Links to an external/public URL 
  - Although technically it's possible to capture this all with local content and not even need a live API which would be great for offline. Your output would look something more like this:

```
> [!Bible] [[Genesis 1|Genesis 1:1-3]]
>  <sup> [[Genesis 1.1|1]] </sup>![[Genesis 1.1^verse]] <sup> [[Genesis 1.2|2]]  </sup>![[Genesis 1.2^verse]]<sup> [[Genesis 1.3|3]]  </sup>![[Genesis 1.3^verse]]
```

2. Easily jump to the exact verse note instead of the starting verse and navigating
3. My note now links to all verse notes which appear in the backlinks like so
<img width="349" alt="image" src="https://github.com/tim-hub/obsidian-bible-reference/assets/1228550/f1b6dfb1-9bab-4bcd-9463-7399876e6bd0">

If I knew how to do this myself I totally would. I just started using your plugin today and I'm very impressed with what it already does but I think this would be a great enhancement for local personal bible management.

**Comment by joey-kilgore** [comment]
@Xtremefaith this is a very interesting setup! But I think this will be something that we should go about solving this after #99 is merged, and then we can look forward to custom formatting options as @JTravisWatson @brennenputh have suggested earlier in the thread.

Personally, I don't actually keep the text itself within my vault, since I find having all of the text already in the vault makes it too cluttered for what I like to do with the graph. I like to just have book and chapter notes, and everything links to those where I can keep notes on those verses as a whole (like historical references/additional commentary info).

I think with all this in mind, to give people the full control, we should take all of the options available and give a text box for custom formatting for each. We can have defaults for them, but allow people to enter their own at all 4 levels of tag/link (global, book, chapter, verse).

**Comment by Xtremefaith** [comment]
@joey-kilgore I understand your preference and I'm sure you're not alone based on other members of the community I've spoken with. Coming from some in-depth tools in the past I really appreciate individual verse notes, this is my way of possessing a digital bible that I can really mark up in the margins sort-a-speak. 

As for my graphview I just filter out the Bible directory and it no longer appears. Alternatively if there was a way to easily reference a separate vault from my PKM vault I would even opt to keep the bible in it's own vault to reduce the file count. It's definitely a tradeoff and I suppose the jury is still out since I only begun doing it this way this year. That said, so far I really like it. This Sunday I used your tool to generate the scripture block and then I inserted by references into the verse numbers and got the results I was expecting. Now if this worked without the extra editing that would be awesome.

<img width="827" alt="image" src="https://github.com/tim-hub/obsidian-bible-reference/assets/1228550/712defaa-b251-4408-868e-2385c8c64e9d">

I do believe that with the writing options/settings in place as you're suggesting, the write output could be made for every users preference.

**Comment by minermaniac447** [comment]
Just my two cents - I feel like it makes more sense to have a _single_ tags field, where we put as many or as few tags as we want (or, alternatively, a list of tags that's able to be whittled down or expanded), rather than enforcing one tag per level. For example, I use a single (nested) tag for book and chapter, don't find the verse level practical, and don't find the overall "#bible" level necessary. Thus, I'd only want to use the one tag. Others may have a use case where they want to use multiple, for example, book-level tags (maybe they have multilingual bible notes? not sure).

Basically, I feel locking into 4 categories, one tag each, is rather limiting.

**Comment by tim-hub** [comment]
> Just my two cents - I feel like it makes more sense to have a _single_ tags field, where we put as many or as few tags as we want (or, alternatively, a list of tags that's able to be whittled down or expanded), rather than enforcing one tag per level. For example, I use a single (nested) tag for book and chapter, don't find the verse level practical, and don't find the overall "#bible" level necessary. Thus, I'd only want to use the one tag. Others may have a use case where they want to use multiple, for example, book-level tags (maybe they have multilingual bible notes? not sure).
> 
> Basically, I feel locking into 4 categories, one tag each, is rather limiting.

Makes sense to me too.

**Comment by joey-kilgore** [comment]
I think all of this is makes for some really good suggestions. I think the first iteration of the settings UI was a bit rough, but I've always been someone who would rather get a rough piece of UI out and running and iterating than just keep it in development.

I think moving to a single tag option in the settings and allowing for people to put in whatever custom tags is a great idea, and I'm just now seeing #109 which looks like it also might address some of the comments here, and @brennenputh and @tim-hub  seem to be doing a great job with it. I'm wondering if it might be worth trying to get some sort of overall community response on how they would like/already use tags and backlinks in their notes (maybe like a google form) so that we can get an idea for how flexible we should aim to make these tags/backlinks. Or we can just aim to make them as flexible as possible, but as we make them more flexible it will become more cumbersome for users to configure (noting we will likely need more documentation for how to handle these from a user point of view).

**Comment by micahmorgan1** [comment]
I think a templating system is the way to go.  I personally would like nested tags, and to be able to tag each verse independently in a referenced passage.  I think as noted above though there are too many permutations of preferences to appease everyone, so some kind of template would be great.  It could possibly even include the callout feature.   I understand there is a lot of added complexity added there, as referenced in #109 so I understand the reasoning behind rejecting the pull request, but I think the idea is solid.

**Comment by JPT62089** [comment]
Yup, I personally would prefer verses to be tagged as something like #bible-reference/{book}/{chapter} and spaces be replaced with dashes (e.g., #bible-reference/1-john/1). I usually do something similar for sermon notes so that I can go back and find my notes from certain chapters. imo verse tagging would be too much but I'm sure there are those who would love it.

**Comment by tim-hub** [comment]
Hi here, Thanks for all of your contribution here.

I moved this issue to discussion, because it is more like an idea brainstorming than an issue here for me now. :) 

new ideas are always welcomed https://github.com/tim-hub/obsidian-bible-reference/discussions/127

**Comment by Xtremefaith** [comment]
I'm trying to understand the value of tagging that many seem to be supportive of in this thread and I suppose a key difference from my use of verse notes is that tagging allows for connecting thoughts/references to an **external bible** whereas verse notes allows for connecting them to an **internal bible**. From that perspective I get why some may view it as too much or unnecessary, but personally I don't see what is so difficult about the verse note option, especially since it has less complexity than tagging. 

All that needs done is to provide an option to convert the `<sup> # </sup>` references to links markdown links of the references verse (ie. `<sup> [[2 Timothy 2.14|14]] </sup>`. Even if it's not a feature everyone uses it's a pretty powerful one considering the ability to easily link your note to every verse it's associated with. Also, this feature wouldn't require lengthy inquiry processes to determine all the possible options either. 

This would provide a great capability for those managing an internal bible while the tagging feature can continue being developed for those that prefer an external bible. From that perspective I don't see these as competing features.

**Reply by tim-hub** [reply]
Thanks for sharing, this is very valuable of how the project goes in the future

**Comment by tim-hub** [comment]
Hi here,

To explain why I removed the outgoing links of book names 


there are 2 issues here:

    The UI, it makes the rendering logic more and more complicated, some users prefer simpler UI, others prefer more function, this is very hard to balance.

    The Link format, Obsidian supports 2 types, dependents on user choice.

    [[]] (by default)
    []()
    We need a proper way to handle the link.

https://github.com/tim-hub/obsidian-bible-reference/pull/119#issuecomment-1705014684

But it seems like users want it back, will work on it to resolve the issues to get it back.

**Reply by brennenputh** [reply]
Aka: Templates 😉
Only true way to allow everyone to be happy with the formatting.
I'll get back to addressing the review stuff this coming weekend

**Reply by tim-hub** [reply]
😆 

BTW, I just think maybe building a React component to combine 

- the logic you did in PR https://github.com/tim-hub/obsidian-bible-reference/pull/109
- And add a `preview` of what it will look like. 

The only concern is I am not sure whether the React component will work on settings. I need to try first


- ref: https://docs.obsidian.md/Plugins/Getting+started/Use+React+in+your+plugin

**Comment by tim-hub** [comment]
Just an update in 2.3.0 

https://github.com/tim-hub/obsidian-bible-reference/releases/tag/2.3.0

The original idea (Thanks @joey-kilgore )  of Book Linking and Chapter Linking is back.

**Reply by cjulin6651** [reply]
Testing this out this morning (I'm excited you're bringing this back! Thank you!) and I'm still not seeing the book/chapter links. I have updated the plugin. I've tested on multiple passages. 
<img width="707" alt="image" src="https://github.com/tim-hub/obsidian-bible-reference/assets/38538687/9f758a44-1a83-448b-927c-36d340eaf76b">

**Reply by tim-hub** [reply]
Hi @cjulin6651 

What is your setting?

There is a change for the setting ui. Sorry forgot to mention

![image](https://github.com/tim-hub/obsidian-bible-reference/assets/2884320/11cc5ece-e084-44ff-acb4-69e10b7028d1)

**Reply by cjulin6651** [reply]
All the settings that make sense to me to be on are on. 
<img width="812" alt="image" src="https://github.com/tim-hub/obsidian-bible-reference/assets/38538687/cd0326c2-fb08-480b-a283-88e13712ad18">

**Reply by tim-hub** [reply]
wow, let me double check

**Reply by cjulin6651** [reply]
I'm on a macbook. I tried restarting the program multiple times and the computer. No difference. I also tried it on my iphone. Same results.

**Reply by tim-hub** [reply]
@cjulin6651 can you try enable the debugger? 

![image](https://github.com/tim-hub/obsidian-bible-reference/assets/2884320/506f3048-0501-4570-942b-bcf41042b484)


somehow I could not reproduce the issue

**Reply by cjulin6651** [reply]
This what you're looking for? 
<img width="528" alt="image" src="https://github.com/tim-hub/obsidian-bible-reference/assets/38538687/77a2efa2-b0f2-435a-a507-d2a1304b4eff">

**Reply by tim-hub** [reply]
Can you try type --John1:1, see what logs are?

**Reply by cjulin6651** [reply]
Forgive me, this is a little outside of what I'm familiar with...I typed in John 1:1 and see no change in the log from the above screenshot. I then entered it into the note and saw no change. But it is likely that I am looking in the wrong spot. I've never used this debugger before.
<img width="1599" alt="image" src="https://github.com/tim-hub/obsidian-bible-reference/assets/38538687/fe7a4372-ca61-46b5-bff1-f56a1f1ea89d">

**Reply by tim-hub** [reply]
I think I know why.

Can you try --john1:1

I think I forgot to update the verse lookup command.

**Reply by cjulin6651** [reply]
<img width="1908" alt="image" src="https://github.com/tim-hub/obsidian-bible-reference/assets/38538687/1c3ec41b-1dad-4913-a97e-1e52bf2c1365">
That changed things

**Reply by tim-hub** [reply]
@cjulin6651 can you try --john1:1 in your editor, instead of the pop-up window/verse lookup command? 


![image](https://github.com/tim-hub/obsidian-bible-reference/assets/2884320/8ec34837-bee2-439c-9e89-1238fed89253)

**Reply by tim-hub** [reply]
BTW, can you confirm the version you use again? 

Because I fround even using the verse lookup command, it still works on my side.

![image](https://github.com/tim-hub/obsidian-bible-reference/assets/2884320/295aad1e-a1d3-43d5-be29-4c6e522c3768)

**Reply by cjulin6651** [reply]
<img width="1907" alt="image" src="https://github.com/tim-hub/obsidian-bible-reference/assets/38538687/5ca71114-74c2-4732-97a1-2b4944610530">
NIV is my version.

**Reply by tim-hub** [reply]
I mean the Plugin version and the Obsidian version. 
<img width="199" alt="image" src="https://github.com/tim-hub/obsidian-bible-reference/assets/2884320/9a0c0e8d-455b-4b02-bcf9-c308e2c6e33f">

**Comment by cjulin6651** [comment]
I have these two as well, but they don't seem to be for your plugin? 
<img width="538" alt="image" src="https://github.com/tim-hub/obsidian-bible-reference/assets/38538687/4ddf952f-06ff-4ca6-be69-50c75d8c4c4c">
<img width="531" alt="image" src="https://github.com/tim-hub/obsidian-bible-reference/assets/38538687/01a7efd3-ceb1-43bf-86d9-f708ce2a9068">
